### PR TITLE
Make it illegal to refer peer works as futures

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
$subject.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/18024

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 1]
```diff
@@ -230,6 +230,9 @@ error.invalid.worker.join.result.type=\
 error.invalid.worker.timeout.result.type=\
    invalid worker timeout result type, expected 'map'
 
+error.invalid.worker.reference=\
+   invalid worker reference ''{0}''
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: It is not invalid, but we don't support it yet. Maybe we can reword this, saying. "worker reference not supported here yet."

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -602,6 +603,20 @@ public void visit(BLangSimpleVariable varNode) {
...
+        transferForkFlag(varNode);
+    }
+
+    private void transferForkFlag(BLangSimpleVariable varNode) {
+        // Transfer FORK flag to workers future value.
+        if (varNode.expr != null && varNode.expr.getKind() == NodeKind.INVOCATION
+                && varNode.flagSet.contains(Flag.WORKER)) {
+
+            BLangInvocation expr = (BLangInvocation) varNode.expr;
+            if (expr.name.value.startsWith("0") && (expr.symbol.flags & Flags.FORKED) == Flags.FORKED) {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: This validation is not scalable. let's create an issue to track this and clean up the implementation.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 3]
```diff
@@ -1276,6 +1276,31 @@ public void visit(BLangSimpleVariable varNode) {
         }
     }
 
+    private void checkWorkerPeerWorkerUsageInsideWorker(DiagnosticPos pos, BSymbol symbol, SymbolEnv env) {
+        if ((symbol.flags & Flags.WORKER) == Flags.WORKER) {
+            // Current location is a worker lambda
+            // And refering symbol is in a toplevel function.
+            boolean isInFunctionTopLevel = symbol.owner != null
+                    && symbol.owner.owner != null
+                    && symbol.owner.owner.getKind() == SymbolKind.PACKAGE;
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: This logic may not work for Object functions.

Comment (hasithaa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Let's fix this in a sperate PR.

**Comment by hasithaa** [comment]
Please fix review suggestions in seperate PR

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 1 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: unsupported worker reference ''{0}'' ?

Comment (hasithaa) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: +1

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 2 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: https://github.com/ballerina-platform/ballerina-lang/issues/18680 created to track this

**Comment by rdhananjaya** [comment]
https://github.com/ballerina-platform/ballerina-lang/pull/18681 add suggested changes

**Pull request merged**